src/actor/calc.rs

```diff
@@ -0,0 +1,301 @@
+fn critical_hit(sim: &SimState, crit: i64, crit_percent_override: Option<&i64>) -> i64 {
+    if !is_crit(sim, crit, crit_percent_override) {
+        return 1000;
+    }
+    critical_hit_damage(crit)
+}
+
+// P(DHR)
+// https://www.akhmorning.com/allagan-studies/how-to-be-a-math-wizard/shadowbringers/parameters/#pdhr
+fn direct_hit_rate(dhr: i64) -> f64 {
+    // ⌊ 550 · ( DHR - LevelModLv, SUB )/ LevelModLv, DIV ⌋ / 10
+    floor(
+        550.0 * (dhr as f64 - (lookup::level_modifiers(lookup::LevelColumn::SUB)) as f64)
+            / (lookup::level_modifiers(lookup::LevelColumn::DIV) as f64),
+        0,
+    ) / 10.0
+}
+
+fn is_direct(sim: &SimState, dhr: i64) -> bool {
+    let roll = sim.rng.random();
+    let probability = direct_hit_rate(dhr) / 100.0;
+    roll < probability
+}
+
+fn direct_hit(sim: &SimState, crit: i64) -> i64 {
+    if is_direct(sim, crit) {
+        125
+    } else {
+        100
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::sim::SimRng;
+    pub struct FakeRng {
+        random_value: f64,
+        random_from_range_value: i64,
+    }
+
+    impl SimRng for FakeRng {
+        fn random(&self) -> f64 {
+            self.random_value
+        }
+        fn random_from_range(&self, _low_inclusive: i64, _high_exclusive: i64) -> i64 {
+            self.random_from_range_value
+        }
+    }
+
+    #[test]
+    fn test_direct_hit_rate() {
+        assert_eq!(66.6, direct_hit_rate(4376));
+        assert_eq!(0.0, direct_hit_rate(380));
+        assert_eq!(31.5, direct_hit_rate(2270));
+    }
+
+    #[test]
+    fn test_critical_hit_rate() {
+        assert_eq!(29.2, critical_hit_rate(4373));
+        assert_eq!(13.5, critical_hit_rate(1783));
+        assert_eq!(5.0, critical_hit_rate(380));
+    }
+
+    #[test]
+    fn test_determination() {
+        assert_eq!(1000, determination(340));
+        assert_eq!(1039, determination(1330));
+        assert_eq!(1157, determination(4326));
+    }
+
+    #[test]
+    fn test_tenacity() {
+        assert_eq!(1000, tenacity(380));
+        assert_eq!(1079, tenacity(2987));
+        assert_eq!(1121, tenacity(4373));
+    }
+
+    #[test]
+    fn test_critical_hit_damage() {
+        assert_eq!(1642, critical_hit_damage(4373));
+        assert_eq!(1485, critical_hit_damage(1783));
+        assert_eq!(1400, critical_hit_damage(380));
+    }
+
+    #[test]
+    fn test_is_crit() {
+        let sim = SimState::new(FakeRng {
+            random_value: 0.5,
+            random_from_range_value: 100,
+        });
+
+        assert_eq!(false, is_crit(&sim, 0, None));
+    }
+
+    #[test]
+    fn test_is_crit_with_override() {
+        let sim = SimState::new(FakeRng {
+            random_value: 0.5,
+            random_from_range_value: 100,
+        });
+
+        assert_eq!(true, is_crit(&sim, 0, Some::<&i64>(&51)));
+    }
+
+    /*
+    fast blade potency 200

```

Comment:
The purpose of this comment block is very unclear.

Reply:
Oops, it's just a combat log. Should be deleted.